api/services/Reddit.js

```diff
@@ -17,137 +17,120 @@ exports.refreshToken = function (refreshToken, error, callback) {
       "Content-Type": "application/x-www-form-urlencoded",
       "Content-Length": data.length
     }
-  }, function(err, response, body){
-    if (body && body.access_token) {
-      callback(body.access_token);
-    } else {
-      console.log("Error retrieving token.");
-      error("Error retrieving token");
-    }
   });
+  if (body && body.access_token) {
+    return body.access_token;
+  } else {
+    throw "Error retrieving token";
+  }
 };
 
-var makeRequest = function (refreshToken, requestType, url, data, rateLimitRemainingThreshold, callback) {
-  exports.refreshToken(refreshToken, callback, function (token) {
-    if (left < rateLimitRemainingThreshold && moment().before(resetTime)) {
-      return callback("Rate limited");
-    }
-    var parseResponse = function (err, response, body) {
-      updateRateLimits(response);
-      var bodyJson;
-      try {
-        bodyJson = JSON.parse(body);
-      } catch (error) {
-        console.log("Error with parsing: " + body);
-        return callback("Error with parsing: " + body);
-      }
-      var callback_error;
-      if (err) {
-        console.log(err);
-        callback_error = err;
-      } else if (response.statusCode != 200) {
-        console.log('Reddit error: ' + requestType + ' request sent to ' + url + ' returned ' + response.statusCode +
-        ' - ' + response.statusMessage + '\nForm data sent: ' + JSON.stringify(data));
-        callback_error = response.statusMessage;
-      }
-      callback(callback_error, bodyJson);
-    };
-    var headers = {Authorization: "bearer " + token, "User-Agent": userAgent};
-    if (requestType === 'POST') {
-      request.post({url: url, formData: data, headers: headers}, parseResponse);
-    } else if (requestType === 'GET') {
-      request.get({url: url, headers: headers}, parseResponse);
-    } else {
-      callback('Invalid makeRequest() call');
-    }
-  });
+// userToken can either be a user's refresh token, or a Promise for an access token.
+// The latter would be used to make batch api calls without needing to request a new access token each time.
+var makeRequest = async function (userToken, requestType, url, data, rateLimitRemainingThreshold) {
+  let token = userToken instanceof Promise ? await userToken : await exports.refreshToken(userToken);
```

Comment:
makeRequest() can now accept either a refreshToken (as it did before) or a Promise which will return an access token. So if we want to do multiple api calls simultaneously, we can first do something like

    var access_token_promise = Reddit.refreshToken(user.redToken);
    // (returns a promise since Reddit.refreshToken is async now)

...and then just pass in `access_token_promise` instead of `user.redToken` to the following api calls. The effect of this is that `Reddit.refreshToken` only ends up getting called once -- we don't waste our ratelimit on getting access token when we already have a valid one.